core/src/main/java/org/elasticsearch/tasks/CancellableTask.java

```diff
@@ -0,0 +1,128 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.inject.Provider;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A task that can be canceled
+ */
+public class CancellableTask extends Task {
+
+    private final boolean hasChildren;
+
+    private static final Tuple<String, CancellationListener> TASK_FINISHED_MARKER = new Tuple<>("task finished", null);
+
+    private static ThreadLocal<CancellableTask> current = new ThreadLocal<>();
+
+    public static void setCurrent(CancellableTask value) {
+        current.set(value);
+    }
+
+    public static void removeCurrent() {
+        current.remove();
+    }
+
+    public static void checkCurrent() {
+        current.get().check();
+    }
+
+    private AtomicReference<Tuple<String, CancellationListener>> cancellationReasonAndListener = new AtomicReference<>();
+
+    public CancellableTask(long id, String type, String action, Provider<String> description, boolean hasChildren) {
+        super(id, type, action, description);
+        this.hasChildren = hasChildren;
+    }
+
+    public CancellableTask(long id, String type, String action, Provider<String> description, String parentNode, long parentId, boolean hasChildren) {
+        super(id, type, action, description, parentNode, parentId);
+        this.hasChildren = hasChildren;
+    }
+
+    public boolean hasChildren() {
+        return hasChildren;
+    }
+
+    public final boolean cancel(String reason, CancellationListener listener) {
+        if (cancellationReasonAndListener.compareAndSet(null, new Tuple<>(reason, listener))) {
+            try {
+                doCancel(reason);
+            } catch (Throwable ex) {
+                // What to do?
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public final boolean cancel(String reason) {
+        return cancel(reason, null);
+    }
+
+    protected void doCancel(String reason) {
+        // Action can override this method to be notified when the task is cancelled
+    }
+
+    /**
+     * Returns true if this task should be automatically cancelled if the coordinating node that
+     * requested this task left the cluster.
+     */
+    public boolean cancelOnParentLeaving() {
+        return true;
+    }
+
+    public boolean isCancelled() {
+        return cancellationReasonAndListener.get() != null;
+    }
+
+    public String getCancellationReason() {
+        Tuple<String, CancellationListener> reasonAndListener = cancellationReasonAndListener.get();
+        if (reasonAndListener == null) {
+            return null;
+        }
+        return reasonAndListener.v1();
+    }
+
+    public void check() {
```

Comment:
can we rename this to `boolean isCanceled()` and then instead of the exception just return a boolean? I think it would be more intuitive and we really don't need yet another exception